GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,260 @@
+        /// <param name="parentDirectory">Paths will only be categorized if they are descendants of the parentDirectory</param>
+        private int CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, string parentDirectory)
+        {
+            int count = 0;
+            foreach (string path in paths)
+            {
+                // Only categorize if descendent of the parentDirectory
+                if (path.StartsWith(parentDirectory))
+                {
+                    count++;
+
+                    // If the path is to the parentDirectory, ignore it to avoid adding string.Empty to the data structures
+                    if (!parentDirectory.Equals(path, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Trim the path to parent directory
+                        string topDir = this.ParseTopDirectory(this.TrimDirectoryFromPath(path, parentDirectory));
+                        this.IncreaseDictionaryCounterByKey(directoryTracking, topDir);
+                    }
+                }
+            }
+
+            return count;
+        }
+
+        /// <summary>
+        /// Trim the relative path to a directory from the front of a specified path which is its child
+        /// </summary>
+        /// <remarks>Precondition: Only even run in the context of TrimDirectoryFromPath(child, parentDirectory)</remarks>
+        /// <param name="path">The path being trimmed</param>
+        /// <param name="directoryTarget">The directory target whose path to trim from the path</param>
+        /// <returns>The newly formatted path with the directory trimmed</returns>
+        private string TrimDirectoryFromPath(string path, string directoryTarget)
+        {
+            return path.Substring(directoryTarget.Length);
+        }
+
+        private void IncreaseDictionaryCounterByKey(Dictionary<string, int> countingDictionary, string key)
+        {
+            if (!countingDictionary.TryGetValue(key, out int count))
+            {
+                count = 0;
+            }
+
+            countingDictionary[key] = ++count;
+        }
+
+        private decimal CalculateHealthMetric(int hydratedFileCount, int totalFileCount)
+        {
+            if (totalFileCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)hydratedFileCount / (decimal)totalFileCount;
+        }
+
+        public class GVFSEnlistmentPathData
+        {
+            public readonly List<string> GitFolderPaths;
+            public readonly List<string> GitFilePaths;
+            public readonly List<string> PlaceholderFolderPaths;
+            public readonly List<string> PlaceholderFilePaths;
+            public readonly List<string> ModifiedFolderPaths;
+            public readonly List<string> ModifiedFilePaths;
+
+            public GVFSEnlistmentPathData(
+            List<string> gitFolderPaths,

```

Comment:
style nit: should these be indented?

Reply:
Yeah I think indentation looks better there!